cpp/src/arrow/python/helpers.cc

```diff
@@ -193,6 +195,32 @@ Status UInt64FromPythonInt(PyObject* obj, uint64_t* out) {
   return Status::OK();
 }
 
+bool PyFloat_isnan(const PyObject* obj) {
+  return PyFloat_Check(obj) && std::isnan(PyFloat_AS_DOUBLE(obj));
+}
+
+bool PyDecimal_Check(const PyObject* obj) {
+  // TODO(phillipc): Is this expensive? Can we make it a global?
+  OwnedRef Decimal;
+  {
+    OwnedRef decimal;
```

Comment:
@wesm @pitrou is there any reason we shouldn't make these `static` global variables, so we don't have to call these APIs all over the place and inside of loops?

Reply:
IMO it's fine to make it a static global (or a singleton, etc.), as long as we don't want to support subinterpreters perhaps.

Reply:
By the way the nested block above doesn't seem needed.

Reply:
Yep, okay I'll refactor and make those static

Reply:
Hm so it appears that string interning of module names doesn't play well with C++ globals. I'll leave these as they are for now.

Reply:
Right, you probably can't use `OwnedRef` here since the destructor would trigger too late. But you could have a `PyObject*`. The only downside is that it would make the object eternal.